src/database-postgresql.cpp

```diff
@@ -0,0 +1,380 @@
+/*
+Copyright (C) 2014 jjb, James Dornan <james at catch22 dot com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "config.h"
+
+#if USE_POSTGRESQL
+
+#include "database-postgresql.h"
+#include "log.h"
+#include "filesys.h"
+
+Database_PostgreSQL::Database_PostgreSQL(Settings &conf)
+{
+	m_conf = conf;
+
+	if (PQisthreadsafe() < 1)
+		infostream << "Postgres backend is not thread safe." << std::endl;
+
+	connectString();
+	dbconnect();
+	prepare();
+}
+
+void Database_PostgreSQL::verifyDatabase()
+{
+	if (PQstatus(m_conn) == CONNECTION_BAD) {
+		PQreset(m_conn);
+			
+		if (ping())
+			return;
+		else
+			throw BaseException(PQerrorMessage(m_conn));
+	}
+
+	return;
+}
+
+void Database_PostgreSQL::prepare()
+{
+	Oid writeFormats[]  = {20, 17};
+	Oid updateFormats[] = {17, 20};
+	Oid readFormats[]   = {20};
+
+	m_res = PQprepare(m_conn, "write_block",
+			"INSERT INTO blocks (pos, data) VALUES ($1::bigint, $2::bytea)", 2,
+			(const Oid *)writeFormats);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+		throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "update_block",
+			"UPDATE blocks SET data = $1::bytea WHERE pos = $2::bigint", 2,
+			(const Oid *)updateFormats);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+		throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "read_block",
+			"SELECT data FROM blocks WHERE pos = $1::bigint", 1,
+			(const Oid *)readFormats);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+		throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "delete_block",
+			"DELETE FROM blocks WHERE pos = $1::bigint", 1,
+			(const Oid *)readFormats);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+		throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "read_pos", "SELECT pos FROM blocks", 0, NULL);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+		throw BaseException(PQresultErrorMessage(m_res));
+
+	PQclear(m_res);
+
+	infostream << "Postgres: SQL statements prepared." << std::endl;
+}
+
+void Database_PostgreSQL::dbconnect()
+{
+	m_conn = PQconnectdb(m_conninfo.c_str());
+
+	if (PQstatus(m_conn) != CONNECTION_OK) 
```

Comment:
It's not removed yet. Check in your editor for an automatical removal of railing whitespaces and tabs.